### DISCUSSION TITLE
Allow lit-html-like templates as worthy, yet build-free JSX replacement

### DISCUSSION THREAD

**Discussion by cloudspeech** [vanjs-org/van#10] [Ideas]
VanJS' HTML-markup-as-nested-JS-function-calls approach to templating is going to be a hard sell to the wider community, especially people that have seen React. They just want something like JSX.

Lit-html has a very nice syntax that comes very close to JSX, yet only uses browser-native means. See https://lit.dev/docs/templates/overview/.

I humbly suggest to not reinvent the wheel but provide an **add-on** to VanJS that supports unchanged lit-html template syntax (properties, attributes, events at minimum, perhaps also ref). This way the vanJS base size can be kept minimal for purists. 

I expect the add-on's size to likewise be very small (using DOMParser).

What do you think?

**Comment by artydev** [comment]
Hy Cloudspeech,

I personnaly love the tiny size of VanJS and would'nt want to see it growing...
There are many other librairies which can offer what you want.

Is this code really hard  to read for React users  ? :  

[Demo](https://flems.io/#0=N4IgtglgJlA2CmIBcA2ADAOgKwHYA0IAZhAgM7IDaoAdgIZiJIgYAWALmLCAQMYD21NvEHIQAHigQAbgAIesWqVIBeADog2ENgnUA+VdRkyAarWoApAMoy2LeDICMGAEwBrAEIyAqgEkZhACd6eAB3PlcDMQB6SSl9akjYmWg1EFoAB3S9AyMJaWSoVM1teGzDGWjY+MrpeINuEFJ4BB5NAXImAGYkNBAAXzwaYNEpMwwAK3JeASERJgQ2GXg8NjxqPFhlAHkAI3H4VrwA7b2DtgwAc3g2AAUAvjYHgE90+C3CPBUACmXV9dgAJTKXQ-AD8oK+TTYABUIAw+ABXNhff4AtahGSWa4AgEYWgwL5sNGFYBXRajWBfAHAALXBEBQy2CCkDBsAZQmQUn7UhYyajKJmkDbKaisgDc8AAhMplLAAGRylEYTYy2CgtjKUiEtZ4HgApDwGUi5UK85QZrXeAotGijXLUV8DCEPgBACitB4LC+hOByOWgJxAIGAmooW5wEFGEd6QRpC98CDweUAS+UDREGUht08Aw1D45uhL3goPgSFDIRk0PgAA8-WiEcofngMC2icDCU6FMiHFE0LjnW6PV6fdm8ZlhFAAMIsEhQL4ptsyqCgkLan7AhP6iCEwNohN4WgijF3PjVp7e7Ut3NA3TABYUIVXiAAXWTKIoaGf4L7Rr4oOoSBUAMV7UM+B7KFAfA8AiDCCBgPC0rQQiuggsHIkSYq0mw9KGLAGDCGwAQQPAWqkP2Lrup6F4UH8z43ne1wyJstA0c+0qZiWwIsWwr6llmtAYFCACCbCERAOxIlaqwJph1pGsuq7rOuIKUvutCBkgKbUECMp-quIEYOa6RCt6V4Ju2lKioQXxmTiB4aZZQZ2QiXy0M2LYQE5gxkkg2rab6GA7BA1BzvaXkAPqZhuGAZjmEBOiQQgpspOYIqCMWkJOIZnPAaZojwjb0by-IUFePFina6IVliyJWV2ACyGRfOFOJOhRQ7ehqt4LMA4VlngCJILAeAXJpfQCp8b4GWATUpaydnDdNs1ZjmAYAmKpDsfKiqbRxeG0ukCg8PAADqWheucDbbmR+pfPt8BgHwUhWmil2ZjifR2VZ7VUXNmw5kSQZ4CEjZlUVjGbOcpCwBAx1fGgeAALQOGixxsDRyrCBctjI2BhQpleeEzek3orfNOJihA1mFO9WE4RQz5iryfDKL1Q0DUg25piNY2YXSDJMW1g6-V1wD8lq6yrN6N5KgOlFei1azvWiDjwJ0r0xRgMZxl8fBOWijoIn0ErVukLqLOahC0AisBsMA+JQEgCIrLQFykEgbmkGwSF8e2vXhek9yPOFfVQCsBobBHECAWBfAx59eBBSFSAhMbDRNC0bTUB0IBoEgDgOAAnIX-SDCAdAMKIExTCA-CCARohwmbATkv49xgDI6gYFEoyipM6gGAYddezIwCkOkZh4DILAOFPsRT3CFwyH0MjKJyrKu6Qg-UMPiwO6vMhfDsigJqvuiH80RJn+vDtHyfU+XwC2+7zYtABGSB+QdBaGXNcKEPQRdwTwfBznUBkLIIAn473aIsYoCBP5QRggRX+bB-5oSASAr46g4GlEgc-GBMgvZPAQOFWM8AAgB1hosNeAADHIMgdgunNAEfO6RqyEL4DDKAMgXRmCuGKehE8YDBVGjITobCBHlBmu-YK+c0BsJoZIgwhAEQ7yzt4JoARD5kICACUe9C6YCy+EYTu5QTEyCiFEDR5CZAhCCOOAI9DzFJC+GPNgxDSzqEkOPBQTwkCEAQNWMUMhaAwwuNQRGWgHru2OvXAIYoQjQFsEgQuaAACkAjS5OPMRYqxXhNEyHSFQ+kpQzE5NiK4ohCAkDqGkRcYKiMiIXHYHIth6gBg5OyTk5IYALiVPcdUqp8BSGaMoYcQhAQeBO00VrYptJWQsBgjsOgJBPpdJMWidZRhLHWK0RXUp3SjAuK2SY8eZgvg6NzMEBKAQvZT3UJ3SBU9DmmJeUYM51ALkzP2cqRQRITlQO6YCjZBg+jbxUWoiAAgbB8HMKQAQ849HAAMfzQwARq4IqgWChI1AIWtChYYbxh1aBPHyeQ0g2jNFkX0WY2+OCASpmkFgkAAAlMwkEO5ktuTIISNwfDqBxK8owOiWS0lIDbNgLJiZfJscCEJBJvbv2xFgspRguUyt0cCmQejTEr1Mco64VF1DsDYMZJAligghT4GAS5DAogZAgKCMVEqVBYAFfQhZwhCSwvhdpD1tgvVEt8Vysi9D6HpwtPi9oogAAsWAkDOERnnXoAwhiVyYPBJQDQ66zDYKIRhUAng0qMMfHgrgLj3FUY7BhR1XCSKMPwWALokAyCgG-VwvDqD8Poc6QQiMraQFgH4kJRFQmSOxQYVkWh4HIvKEIWsiNQkQHCUgWJSV63+BmIjUgEAABepYZDOFpGADdQjJBdpbZ0LAEjQXbwAMTgNHkK0t5bK3JxkO-Y+XxnBYCwHgBwWAUB4ARpgZwKNT0OxES25w8j2EwbYYe2DiGb3lESVAZJMhf2wY3fOtgi6wkATXeQjddTZEyDQCEpEfBx3bwjZnAlOcYM9H6GBcuwwmD1WMF4UwFhLAhv6EAA)

```js
import v from "./van.js"

const {span, h1, div, img } = v.tags

const add = (base) => (elt) => v.add(base, elt)

const target = document.getElementById("app")
const title = document.getElementById("title")

const style_user_pict = `
  border:1px solid orange;
  padding: 3px;
  margin:10px`;

function User (user) {
  return (   
      // User wrapper
      div ({style:"display:flex; align-items:center;width:90%;"},
        // User picture
        div({style:"margin-right:10px"},     
          img({style:style_user_pict, src:user.picture.thumbnail})
        ),
        // User name
        div (
          span(user.name.first, " "),         
          span(user.name.last)
        )
      )
    )
}

function toJson(r) {
  return r.json()
}

function displayUsers (users) {
   add(title)(div("Random Users API")) 
   users.results.map(user => add(target)(
      User(user))
   )  
}  

fetch("https://randomuser.me/api?results=5")
  .then(toJson)
  .then(displayUsers)
```

And this one ? [Demo](https://dev.to/artydev/simple-stopwatch-in-plain-js-fbk)

```js
import van from "./van.js"

const {button,span, div, pre, h1} = van.tags

const elapsed = van.state("0.00")

const Stopwatch = () => {
  let id;
  const start = () => id = id || setInterval(() =>
    elapsed.val = (Number(elapsed.val) + 0.01).toFixed(2), 10)
  return span(
    button({onclick: start}, "Start"),
    button({onclick: () => (clearInterval(id), id = 0)}, "Stop"),
    button({onclick: () =>
      (clearInterval(id), id = 0, elapsed.val = "0.00")}, "Reset"),
  )
}

const Screendisplay = () =>  h1(elapsed, " s")

van.add(app, div(
  h1("A Simple Stopwatch"),
  div ({class:"stopwatch"},
    div({class:"content"},
      Screendisplay(),
      Stopwatch())      
    )
  )
)
```







VanJS is very orignal like it is.
Using it with templates would make ressembles to any other libraries.

The risk of creating addons is to lost new users, and finish like the Jquery Plugins, with all
the versionning problems, and plethora choices....
But Tao can have a very different vision...

Let us keep VanJS thin (and even thiner if possioble :-)) and performant

Look at HyperApp, it started thin, and now :....

![image](https://github.com/vanjs-org/van/assets/31207473/b33ad7c7-c0bf-47cb-a453-7f7b4bb948ce)

No offense

Regards

**Reply by Tao-VanJS** [reply]
Thanks @cloudspeech for the suggestion, and thanks @artydev for your detailed discussion! Really appreciated it. There are many JS frameworks that use template literals to compose DOM trees. It's actually hard to be sure that if VanJS chooses this route for its API, it can indeed provide extra benefit compared to these alternatives.

API style is a very subjective topic, and it's totally understandable people have their own preferences and opinions over styling choices. To me, I like VanJS's style cuz everything is just plain JavaScript code which means you can get syntax highlighting, auto formatting, linting, jumping to definition, refactoring, etc. from any development environment that you're used to.

That said, if you like VanJS's implementation and the template-literal-based API style, since VanJS is open sourced under MIT license, you can feel free to build something on top of VanJS and perhaps publish a new repo under your ownership. I will be very delighted to see this kind of additions to VanJS "ecosystem".

**Reply by cloudspeech** [reply]
@artydev Yes, the code is really hard to read for React users - did you meet any lately? You underestimate the lock-in that JSX has caused in the wider community. JSX/lit-html templates allows to leverage HTML-like notation, so not only React, but anybody who knows HTML will feel at home.
@Tao-VanJS My motive is to prevent your really nice project from oblivion by an unfortunate choice of syntax. I stressed it should be an add-on/plug-in/extension, which maps to your original style under the hood, so nothing is destroyed and people are given a choice of surface syntax to program in.

**Reply by Tao-VanJS** [reply]
Thanks @cloudspeech for the suggestion! I am currently considering an add-on model where 3rd-party contributors can contribute code via VanJS add-ons. New features of VanJS can be considered to be added as add-ons first. This allows us to see users' reception about the feature. If the users' reception is good, we can consider merging the feature to core `van.js`. To provide a feature as an add-on, we can implement a library that takes `van.js` as a dependency, and exposes the additional APIs. You can implement the library in a subdirectory under `addons/...`. A README.md file can be provided for the instructions of using the add-on. And in VanJS home page, we can have a dedicated section to talk about all VanJS add-ons implemented by 3rd-party contributors. If you're interested in contributing to VanJS with an add-on, we can talk about the details of the logistics.

**Reply by artydev** [reply]
@cloudspeech 

OlivMenier made a nice suggestion : [htm](https://github.com/vanjs-org/van/issues/11#top)

```js
import htm from 'https://unpkg.com/htm?module'
import van from "./van-0.11.10.min.js"

function h(type, props, ...children) {
  const tag = van.tags[type]
  if (props) return tag(props, ...children)
  return tag(...children)
}

const html = htm.bind(h)

const counter = van.state(0)

const app = html`<div>
  ❤️ ${counter}
  <button onclick="${() => ++counter.val}">👍</button>
  <button onclick="${() => --counter.val}">👎</button>
</div>`

document.body.appendChild(app)
```

You can test it here [VanHTM](https://flems.io/#0=N4IgtglgJlA2CmIBcBWAnAOgEwHYA0IAZhAgM7IDaoAdgIZiJIgYAWALmLCAQMYD21NvEHIQAHWoSAPFAgA3AATQAvGJC0ADhrUA+aQHpZcvZOrcQpeAh5sIA8kyxIAjABYQAXzw16jEHNpqDAArcl4BIREmfX0FAEEAVzY+AFo4rWEoBUJ4WjYEgCd4BQkENgUAfQAjWks8StI2Wh4AawVlBQoAXQBuBX6FGIUAZSbW7L4ChUbJiGoAc0qaywUNPlIIW3sJCSHLctoFangAdwVl+DW58sJJhU0NTPrSWjlijSK5OwTSBQFz2rwUbNFoSQgJag2OzUc7weZzAAUAEkACIASgUwAk-QqjRBGA0PxYCOqgLRPSGw1exT4sCyF2xg30EEIChJeNaGAQCzYLAUOgUzgADEKMbyCnwzsczgBRAoSgoItRqcmM0krDoItgAT0efFZyPR7WUHQA5I0CnN5qaMQB+BRQPg8BIMQQYebwNgyhCutgAIW1SKghoxSAUqMZRXyBRh6vgEg8u1iRRmRQUsFq5Qu9R4gjDPFosFgCjACVgtg0CAUtgYpDBEKh-0yCNz5Q6zgxWJhSgNrf5goxzb7KQHkc9hRhJIu7QaYxaBL4GgRaIxAyGKeSadpWQ5oOoidMQwAKiwIL9iFX+NQNo1hGxYNqFD94L8AAZlV8KFrwbUnSZZOZb1oLJ9QUV8r0aT9bimIooASHgrWrFhiiqCE4GKDYAC94AwHZqCGOJYAgWpENIPgGHOBISFsGFSG1MAqlpX5kgUWD4OKXkULQqssJw0pPQUAB5KpghnYTgngGx6gqND4GIY4oHqD4+GSQTWQ6cTJLYd1PQABQlZIdUeNS8LKe4YDiagoGGHhkLghBBOoAAxCACkaGcEX2Z56nBSF6hOWhNgAWVIDFlBMfpPIE217SitgjwgBg+CSBFfJ4fzArYEK0XqaURk9FcMGA4NQtMgTbMCD1rKaIQ61MMy8SEfSVL4Gcu36D1yjUAJYDUZdMVY8cYyQs8MAqHqFC8PD+n2Ep-ELPq5E7Rl+gIoiSIWMDeTPT8WLYngOOQyirJ4iBsNw7t+gamdttIHNCgANRnUgxp6laezZRQAEITQUZ0Cge5bLoGFk2Re7cHsLY0On+wH3oGP6WEq+BqryF8Z2KyzrNslGy3gRyXLctgWyRhYUeBWrvKfDQoDRlFyNK4H+isFZQYRRQTQ6cG6Uh2A0XhgYKrJ1HaowKArE9eBPP5pnpleqGOjkAWXqI29jjcjBoJlZpiWLcL03Z+6AZXd6D36KbTH6NQBGlPq+YG26uTPSJ4A1wlSF1jELYPJNphq+BmpYs97lWDM5j+EStKQn5VgMvg1M8-3A74DFg+CH5yk0mwCTjoy+Pw2JlNUwhE7R5PU9+eAAEcqJ6u9q1aov45L4APAxVDyheN5fmcc5tSEJQYU4ks5jOlHjvQi6zL4ETk5nJuE8agO45l-jO-9mdR7YXn2gFBF2sqCom4qCowyX5O8DVHqwy33nL+7CHCxv6hNjvxkqjmWQFlIMNunvq7nbCFdj-ToXR75tzKuUZI9MwAzg5gKOQGBqB8HFkeXUxR7SKDDHlI88AAAexMlqQPMlkTUjowD1AwFQ2yJAoBFGoGFPejIaF0noZrDMxMkTUHkpsbUaJNaTG1rZEmtDd4OnIkVDIVkADCp46RKmBgvEuLCoBhV+ufOOCh7QfysiIuk9R4EKCWgoMM0DyJ6NUSuf+4iwCr2oGZJo8xfgdDys1PB2oEQIjoAwShVDaAFCcYwzEjIzIUGUhoO6CgqEYBUfQro8844J38U4igQp4kxQUKKaGEdghz3tMk34v9W6+KKgE0gXQQkCXITOR0zpfQxKKGjb08BfReN8DLfo0E2ShJaAY+JoFxIYDvJaF8CJwmhSBgjBqnohBTA6OQigLR4k-Q6DJKyck5jj0wWIhZSy4EmKMTsiR+w4hsDYJaDuUtelGI6SDA0Sj2ZqK5knFe5xP6G0OfrOKszHn1HISbYGLNihswecYzmOTy5vN0dEuQFA1DiwiWoMBbJokItCmI75rt2ZwqIEihEqLLgTL+eRAFCMgXTBmVioh3YzbpgEnMkhCJyElNicIW5bM4wYmKlOQE9QCgYiGLQJIqR0iPCsvUUG05kHlBaRoHUY5owwgKAmeowBOphi1LQeYuV2liMcRgHRwY1ni3kijHVDA0QQPqgJC8sy-SfytM46YYiXqGsdc9A1Drv6a2oliqoYiqhi3IraDAZ5pE2y0ijOxZkEg0zpgzDyQSD5mSFlVCmL44jyloI+DoFBompvJv7cpzDSZpqLTOY1GyFKMi6QiMy-rQJ5WGJ6EmyMRYZqzdqdheQgqaFSr6go9qrKOpXJMgYZlgAVDRcSsAHgZxVHemZaUMD52awbPiqhU7CUYDAH20hAooDyz5jOkpW6Ik7r3TsjAj8+a3P6GzZd5EFArJsXeu5bJH2wJ+pUWSprVEC36OQjARRKzNHgAAdU2MSQN1SOhmLAF404MCVx9HJUBooYA+BvGXPUGDT7Vm-s2VARktLa3XVAgW9tpBM0FGzRibmUAd5cyPQmYhbqFgAHEeDSO1DwBAXCQozmFCKNefs0akCPHwLjbHP4eQJREpNlSbgNhqdutYS5blmRTGWNs2R13yZerupc+6HRHtJR9BE2nyzGh-esv9GIowTlAUpqFX9FgdEnWisM06bGmL4DAyzL4dM5T05CUjdw60CV+KBNFY6ZpFsk1xjGFkrI2TsnjAmrlGil1qol9K0xrEI36l8peEmpM8AEQUIRxJbWuyHW50KzwEvlYrYRhSOUBbsfmFxnjfH4ACcZgjV1XqnEEiJAiLrtzaWOeGl1oNYBWOmHwWsAo5QTVCus8AYq9RHGRKXrhz+uG4RzHqJkA85hLDWC2NeUQzgnC4E8N4EA3i-AhDCCAK8kQ2CiESit8o7BYGEAlLA007A2ARKQDECEGgWjzBieRfQAPbSYfsvAU0EhfuTHKAEGEQOn1qAwPoHHb3lSmDStdhQxI85KQlBEllci6FsuCd2CCUCtVwMCBgXbFA84VO7CC2n6KZtDy1WMwX9PaH0NucL6sov80M6l4tiQrPKccD1qrsAnrdEsDsSr-gEJZkc6CEvBEoo8Iq4eDOAHsBXwyHkBFBQgATckAPB-CgAAkwB9eCFdrSqQHdkgwgEHxiArRVAgA98VgUABqKPXvZkYB6omEAOhAC8G4AWR2pD6H9wIB3fukgB7+JCIioe1AR6CSkFIcfXYJ8LEn1PgA5Hcz9n6gJhM9GB0K+PCtSXR3gNSgrtDxMiyNoQiB4MsLsS0bDdpgzg0BIAAMxCkez4BgogYmkHe59u8ohGJQEfAfXdAS5hICFD0RkGhipWhP0r6gABiS3B8aitHmBKNCYYqiwASPAM-LPaR3DDCcKeEID-h4BPldtCA4CAE4M4M4J4GAs9r4KIJDNQCiEFAADIAAStI2ongQAA)

**Reply by efpage** [reply]
A simple HTML to DOM conversion can be provided like this:
```JS
function html(s) {
    let el = document.createElement('template');
    el.innerHTML = s;
    return el.content;
}
```
I´m just not sure if that plays well with Van states...

**Reply by jfloresremar** [reply]
Hello @artydev, coul'd you please tell me which IDE are you using to keep this format?:

![image](https://github.com/vanjs-org/van/assets/10441071/fc8585e7-a1f8-4f06-8902-d703213d627f)

I use vscode with prettier extension and i don't know how to keep code formatting like your example. I have "format on save" enabled and this is the result (difficult to read)

`export const Home = (app: Elysia) =>
  app.get("/", () =>
    van.html(head(title("SSR")), body(div({ id: "app" }, h1("Titulo"))))
  )
`
Thank you!

**Reply by metaleap** [reply]
> @cloudspeech
> 
> OlivMenier made a nice suggestion : [htm](https://github.com/vanjs-org/van/issues/11#top)

I changed the `h` like so:

```typescript
const html = htm.bind((type: any, props: Record<string, any>, ...children: any[]) => {
  const tag = (typeof type === 'function') ? type : van.tags[type]
  if (props)
    return tag(props, ...children)
  return tag(...children)
})
```

and this lets you do this:

<img width="681" height="587" alt="image" src="https://github.com/user-attachments/assets/7ec706ac-1093-46a1-81ea-b4b52719b464" />

**Comment by yerlaser** [comment]
Very bad suggestion. Almost all other JS frameworks use JSX and it's ugly, too verbose, too much noise vs data.
If you're using some editor plugin, it can save some typing by auto-closing tags, but the noise remains.
Still with many editors syntax highlighting breaks, things like commenting, auto-identation, linting etc. would need more and more plugins leading to growing bloat.
XML/HTML tag syntax is so bad for humans that people started to create alternative markup syntaxes, Markdown, YAML, TOML, JSON etc.
Familiarity with HTML should not be an impediment for JS progress.

**Comment by cqh963852** [comment]
@Tao-VanJS Hi, Can you check our [repo ](https://github.com/idealjs/van-jsx) a jsx-runtime powerd  by VanJS.

I think this implementation nicely solves the problem under discussion

**Reply by Tao-VanJS** [reply]
Thanks! This is nice! Could you add some instructions on how to use the runtime in `README.md` file? I can add a link to the repo from VanJS homepage if it's usable. If you want to port the code to `addons/...` of the VanJS repo, you are very much welcome:-)

**Reply by cqh963852** [reply]
The README.md has been updated. [Create A New Project](https://github.com/idealjs/van-jsx#create-a-new-project) describes how to create a project using [@idealjs/van-jsx](https://www.npmjs.com/package/@idealjs/van-jsx)

**Reply by cqh963852** [reply]
At present, it seems that the `addons/...` is not suitable for placing a van-jsx.

Since it needs package.json to tweak some exports, I don't know how to make sure it's published in the `addons/...` folder as well

**Reply by Tao-VanJS** [reply]
Hi @cqh963852,

`package.json` file can be under a sub-directory of a repo. I did that in many sub-directories of VanJS, like this one: https://github.com/vanjs-org/van/tree/main/components.

I added the link to your project as a community add-on: https://github.com/vanjs-org/van#community-add-ons.

Thank you so much for your contribution!
